crates/symbolicator-service/src/metrics.rs

```diff
@@ -1,17 +1,342 @@
 //! Provides access to the metrics sytem.
 use std::collections::BTreeMap;
+use std::fmt::Write;
 use std::net::ToSocketAddrs;
-use std::sync::OnceLock;
+use std::ops::Deref;
+use std::sync::{Arc, Mutex, OnceLock};
+use std::time::Duration;
+use std::{fmt, thread};
 
-use cadence::{BufferedUdpMetricSink, QueuingMetricSink, StatsdClient};
+use cadence::{BufferedUdpMetricSink, MetricSink, QueuingMetricSink, StatsdClient};
+use crossbeam_utils::CachePadded;
+use rustc_hash::FxHashMap;
+use thread_local::ThreadLocal;
 
-static METRICS_CLIENT: OnceLock<StatsdClient> = OnceLock::new();
+static METRICS_CLIENT: OnceLock<MetricsWrapper> = OnceLock::new();
 
 /// The metrics prelude that is necessary to use the client.
 pub mod prelude {
     pub use cadence::prelude::*;
 }
 
+#[derive(Debug, Clone)]
+struct Sink(Arc<QueuingMetricSink>);
+
+impl MetricSink for Sink {
+    fn emit(&self, metric: &str) -> std::io::Result<usize> {
+        self.0.emit(metric)
+    }
+    fn flush(&self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
+type LocalAggregators = Arc<ThreadLocal<CachePadded<Mutex<LocalAggregator>>>>;
+
+/// The globally configured Metrics, including a `cadence` client, and a local aggregator.
+pub struct MetricsWrapper {
+    /// The raw `cadence` client.
+    statsd_client: StatsdClient,
+
+    /// A thread local aggregator.
+    local_aggregator: LocalAggregators,
+}
+
+impl fmt::Debug for MetricsWrapper {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MetricsWrapper")
+            .field("statsd_client", &self.statsd_client)
+            .field(
+                "local_aggregator",
+                &format_args!("LocalAggregators {{ .. }}"),
+            )
+            .finish()
+    }
+}
+
+/// We are not (yet) aggregating distributions, but keeping every value.
+/// To not overwhelm downstream services, we send them in batches instead of all at once.
+const DISTRIBUTION_BATCH_SIZE: usize = 20;
+
+/// Creates [`LocalAggregators`] and starts a thread that will periodically
+/// send aggregated metrics upstream to the `sink`.
+fn make_aggregator(prefix: &str, formatted_global_tags: String, sink: Sink) -> LocalAggregators {
+    let local_aggregators: LocalAggregators = Default::default();
+
+    let aggregators = Arc::clone(&local_aggregators);
+    let prefix = if prefix.is_empty() {
+        String::new()
+    } else {
+        format!("{}.", prefix.trim_end_matches('.'))
+    };
+
+    let thread_fn = move || {
+        let mut formatted_metric = String::with_capacity(256);
```

Comment:
Why do we reserve 256, can you add a comment? I see that my previous comment is marked as resolve, but I couldn't find the comment.

Reply:
Yes, I added that to the place you were commenting before:
https://github.com/getsentry/symbolicator/pull/1379/files#diff-47987166200c61a727b0e0ef4e81a849a0dd2c78229de06941be16517368a171R273-R275

Now I added that same comment to both places.